### PR TITLE
SourceKitten 0.3.0

### PR DESCRIPTION
**Pull request by jpsim**
This is a work in progress PR to address #17 that marks a major overhaul of SourceKitten focusing on modularity, testability and architecture.

Using dynamic frameworks through [Carthage](https://github.com/Carthage/Carthage) for dependencies and keeping core functionality in SourceKittenFramework should help in the long-term maintainability of this project.

The project structure is heavily inspired by Carthage itself so a HUGE thanks to the people behind that project for their hard work. Shipping command line tools with Swift frameworks is pretty tricky and they've managed to find a decent solution, which I opted to piggy-back on :pig2:.

Steps until we can merge this:

* [x] Write comment docs. Current comments are entirely out of date, so please don't pay attention to them.
* [x] Write more tests. Especially in SwiftDocsTests, ModuleTests and OffsetMapTests.
* [x] Make sure jazzy's integration tests still work and manually test a few other projects.
* [x] Make sure sourcekitten will prioritize frameworks in the same directory over the ones in ~/Library/Frameworks.
* [x] Reintegrate ObjC/Clang proof-of-concept.
* [x] Re-add "parsing File.swift (1/1)" stderr messages.
* [x] Print SourceKittenFramework errors.
* [x] Find work-around for SourceKit inconsistency issue.
* [x] Write proper CHANGELOG entry.

That being said, if you keep in mind that I have yet to complete these steps, I'm very open to review. I'd recommend you mostly ignore the diff and take a more comprehensive look.

### CODE REVIEW

File: Source/SourceKittenFramework/Request.swift [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    /**
+    Sends the request to SourceKit and return the response as an XPCDictionary.
+
+    :returns: SourceKit output as an XPC dictionary
+    */
+    func send() -> XPCDictionary {
+        dispatch_once(&sourceKitInitializationToken) {
+            sourcekitd_initialize(); return
```
Comment (segiddins) [inline Source/SourceKittenFramework/Request.swift]: `return` is unnecessary

Comment (jpsim) [inline Source/SourceKittenFramework/Request.swift]: actually, it kind of is (https://github.com/robb/Cartography/issues/9)

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,20 @@
+## 0.3.0
+
+##### Breaking
+
+* Everything. No, seriously lots has changed in this release and you should
+  consider SourceKitten entirely rewritten. SourceKitten now uses dynamic
+  frameworks for the bulk of its functionality, which means that everything is
+  now much more modular and testable.
```
Comment (segiddins) [inline CHANGELOG.md]: Forgot the double space

File: Source/SourceKittenFramework/File.swift [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+            return nil
+        }
+    }
+
+    public init(contents: NSString) {
+        self.contents = contents
+        lineBreaks = contents.lineBreaks()
+    }
+
+	/**
```
Comment (segiddins) [inline Source/SourceKittenFramework/File.swift]: comment alignment

File: Source/SourceKittenFramework/File.swift [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+        if let parsedDeclaration = parseDeclaration(dictionary) {
+            dictionary[SwiftDocKey.ParsedDeclaration.rawValue] = parsedDeclaration
+        }
+        return dictionary
+    }
+
+    public func furtherProcessDictionary(var dictionary: XPCDictionary, documentedTokenOffsets: [Int], cursorInfoRequest: xpc_object_t) -> XPCDictionary {
+        let offsetMap = generateOffsetMap(documentedTokenOffsets, dictionary: dictionary)
+        for offset in offsetMap.keys.array.reverse() {
```
Comment (segiddins) [inline Source/SourceKittenFramework/File.swift]: why the `reverse`?

File: Source/SourceKittenFramework/File.swift [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+                    fatalError("No mapped offset for \(offset) in \(filename)")
+                }
+            }
+        }
+        return dictionary
+    }
+
+    public func newSubstructure(dictionary: XPCDictionary, cursorInfoRequest: xpc_object_t?) -> XPCArray? {
+        if let substructure = SwiftDocKey.getSubstructure(dictionary) {
+            var newSubstructure = XPCArray()
```
Comment (segiddins) [inline Source/SourceKittenFramework/File.swift]: why the `var`?

Comment (jpsim) [inline Source/SourceKittenFramework/File.swift]: because we append to it a few lines down.

File: Source/SourceKittenFramework/File.swift (hunk 4 above)
Comment (jpsim) [inline Source/SourceKittenFramework/File.swift]: I honestly don't remember. I should have written a comment about that decision when wrote the code, obviously. I'll have to investigate at some point.

**Comment by jpsim** [comment]
I ended up fixing some regressions and now jazzy's integration tests pass.

**Comment by jpsim** [comment]
We could definitely benefit from more tests (especially in SwiftDocsTests, ModuleTests and OffsetMapTests) but IMO is no longer blocking merging this PR, so I checked the box.

**Comment by jpsim** [comment]
Documentation comments have been added for everything in SourceKittenFramework.

**Comment by jpsim** [comment]
We're now printing errors to stderr again. It would be nice to use `Result` to return monadic errors at some point, but that's out of scope for this PR.

**Comment by jpsim** [comment]
It's amazing that without making functional changes to the SourceKit-related code, we've managed to exhaust SourceKit's source text buffer and now the SourceKit test is failing. I'm hopeful we'll be able to find a work-around for this in the next few days so we can finally merge this PR.

**Comment by jpsim** [comment]
The only step left to do in my checklist before this can be merged is to find a work-around for the SourceKit inconsistency issue. That's a big one.

This problem means that with anything but the smallest of projects, SourceKitten will crash after parsing a few files because it exhausts SourceKit's source text buffer, usually printing out the following message:

```
sourcekit: [1:connection-event-handler:7427: 0.0000] Connection interrupt
sourcekit: [1:pingService:7427: 0.0008] pinging service
sourcekit: [1:sourcekitd_send_request_sync:1299: 0.0023] request dropped while restoring service
```

I'm investigating a few possible work-arounds and hope to find something usable soon.

**Comment by jpsim** [comment]
All the checklist items have been addressed. There are still many more points to address, which should be done by opening new GitHub issues.

**Pull request merged**